docs/designs/xref.md

```diff
@@ -0,0 +1,103 @@
+# Xref
+- User can define uid with some properties in markdown, JSON and YAML files. Then output xrefmap.json for others to reference to. 
+- User can reference to uids in markdown, JSON and YAML files, they will be resolved during `docfx build`.
+
+## Define uid
+- User can define uid in YAML header of markdown file
+```
+---
+uid: a
+title: ASP.NET Documentation
+description: Learn how to develop ASP.NET and ASP.NET web applications. Get documentation, example code, tutorials, and more.
+---
+```
+- User can also define uid in JSON
+```json
+{
+  "$schema": "https://raw.githubusercontent.com/dotnet/docfx/v3/schemas/TestData.json",
+  "uid": "a",
+  "title": "ASP.NET Documentation",
+  "description": "Learn how to develop ASP.NET and ASP.NET web applications. Get documentation, example code, tutorials, and more."
+}
+```
+- And define uid in YAML
+```yml
+#YamlMime:TestData
+uid: a
+title: ASP.NET Documentation
+description: Learn how to develop ASP.NET and ASP.NET web applications. Get documentation, example code, tutorials, and more.
+```
+
+## Output xref map
+Docfx will output a JSON file named `xrefmap.json`. In V2, docfx used to output `xrefmap.yml`, it took much longer to be de-serialized compared to JSON format.
+```json
+{
+    "references": [
+        {
+            "uid": "a",
+            "href": "aspnet/a.md",
+            "name": "ASP.NET Documentation",
+            "fullName": "ASP.NET Full Documentation"
+        }
+    ]
+}
+```
+This file will be updated to some central place, and there would be a step to merge all the JSON files for the same share base path [Related issue](https://github.com/dotnet/docfx/issues/3487).
+
+## Create xref map
+There are two parts of xref map, internal and external.
+### Internal xref map
+User can define uids in the same repository and reference to them, and these defined uids will be outputted as `xrefmap.json`.
+
+### External xref map
+User can use uids defined by other repositories as well, just define the urls of `xrefmap.json` in `docfx.yml`.
+```yml
+xref:
+  - http://url1/xrefmap.json
+  - http://url2/xrefmap.json
+```
+During `docfx restore`, all the JSON files will be restored and merged.
+
+## Resolve xref
+### Using `@` to reference a uid in markdown
+  Not supporting `displayProperty` for this case for now, we need to refine the validation pattern of uid firstly.
+```
+Link to @a
+```
+  The file above would be resolved using the restored xref map
+```json
+{"content":"<p>Link to <a href=\"aspnet/a.md\">ASP.NET Documentation</a></p>\n"}
+```
+  - User can also define which property to display for the referenced uid
+
+### Using `xref` to reference a uid in markdown
```

Comment:
some other syntax supported in v2 not mentioned. All of these should be supported?
* more options: [https://dotnet.github.io/docfx/tutorial/links_and_cross_references.html#advanced-more-options-for-cross-reference](https://dotnet.github.io/docfx/tutorial/links_and_cross_references.html#advanced-more-options-for-cross-reference)
* markdown link style xref: [https://dotnet.github.io/docfx/tutorial/links_and_cross_references.html#markdown-link](https://dotnet.github.io/docfx/tutorial/links_and_cross_references.html#markdown-link)

Reply:
what's the behavior if `@a` cannot be resolved?

Reply:
`uid-not-found` warning will be logged if @a cannot be resolved, added to the spec

---
In reply to: [231107690](https://github.com/dotnet/docfx/pull/3578#discussion_r231107690) [](ancestors = 231107690)